process_mets_input/MetsToSchema.py

```diff
@@ -0,0 +1,109 @@
+import xml.etree.ElementTree as ET
+from io import StringIO
+from pathlib import Path
+
+
+class MetsToSchema():
+    def __init__(self, config, descriptive_metadata, structural_metadata, image_data):
+        self.descriptive_xml = self.xml_to_dict(descriptive_metadata)
+        self.structural_xml = self.xml_to_dict(structural_metadata)
+        self.config = config
+        self.image_data = image_data
+
+        self.descriptive_namespaces = dict([
+            node for _, node in ET.iterparse(
+                StringIO(descriptive_metadata), events=['start-ns']
+            )
+        ])
+
+        self.structural_namespaces = dict([
+            node for _, node in ET.iterparse(
+                StringIO(structural_metadata), events=['start-ns']
+            )
+        ])
+
+        self.id = self.descriptive_xml.find('.//dcterms:identifier', self.descriptive_namespaces).text
+        self.base_url = config['manifest-server-base-url'] + "/" + self.id
+        self.default_image = {}
+        self.has_part_items = self.has_part_items()
+        self.main_item = self.main_item()
+        self.errors = []
+
+    def get_json(self):
+        graph = [self.main_item] + self.has_part_items
+        return {
+            "@context": "http://schema.org",
+            "@graph": graph
+        }
+
+    def main_item(self):
+        fieldmap = self.mappings()
+        main = {"@id": self.base_url, "@type": "CreativeWork"}
+        for schema_key, xml_xpath in fieldmap.items():
+            xml_xpath = ".//" + xml_xpath
+            xml_value = self.descriptive_xml.find(xml_xpath, self.descriptive_namespaces)
+            if xml_value is not None and xml_value.text is not None:
+                main[schema_key] = xml_value.text
+
+        main['thumbnail'] = self.default_image
+        main["hasPart"] = []
+        for hasPartItem in self.has_part_items:
+            main["hasPart"].append(hasPartItem["@id"])
+
+        return main
+
+    def xml_to_dict(self, xml):
+        return ET.fromstring(xml)
+
+    def has_part_items(self):
+        xml_xpath = './/mets:structMap/mets:div/mets:div'
+        xml_value = self.structural_xml.findall(xml_xpath, self.structural_namespaces)
+
+        ret = []
+        for index, item in enumerate(xml_value):
+            filename = item.find('.//mets:fptr', self.structural_namespaces).attrib['FILEID'].replace('ID_', '')
+            position = item.attrib['ORDER']
+            name = item.attrib['LABEL']
+            file = Path(filename).stem
+            schemaImage = {
+                "@id": self.base_url + "%2F" + filename,
```

Comment:
The original METS id is of the form ID_2017_007-v0001.jpg.  Once you strip the "ID_", the filename is of the form:  2017_007-v0001.jpg.  The actual object_id should be of the form 2017..007 (in this example).  I'm questioning if the schemaImage should actually be  self.base_url + "%2F" + object_id.

Reply:
That is a good question.  
In the past, we have copied all the images to the image server in the format of 
object_id/image_filename 

Then since the image server is dumb it has to be listed in a URL as 
object_id%2Ffilename 

That is why I am converting it to a file path that is representative of the structure that it will take on the image server so the URLs resolve.   There may be a better way to do this however and if you have ideas  I would certainly hear them

Reply:
Sorry, I was confused.  I was thinking of base_url as "https://marble-manifest-test.libraries.nd.edu" instead of "https://marble-manifest-test.libraries.nd.edu/2017.007".  Since we can have any number of images for an object, of course we do need something more descriptive to identify each specific image... like the filename.  Sorry for the confusion.